pyro/infer/mcmc/util.py

```diff
@@ -153,3 +132,81 @@ def log_prob(self, model_trace):
+    """
+    def __init__(self,
+                 model_trace,
+                 has_enumerable_sites=False,
+                 max_iarange_nesting=float("inf")):
+        self.has_enumerable_sites = has_enumerable_sites
+        self.max_iarange_nesting = max_iarange_nesting
+        # To be populated using the model trace once.
+        self._enum_dims = {}
+        self.ordering = {}
+        self._populate_cache(model_trace)
+
+    def _populate_cache(self, model_trace):
+        """
+        Populate the ordinals (set of ``CondIndepStack`` frames)
+        and enum_dims for each sample site.
+        """
+        if not self.has_enumerable_sites:
+            return
+        if self.max_iarange_nesting == float("inf"):
+            raise ValueError("Finite value required for `max_iarange_nesting` when model "
+                             "has discrete (enumerable) sites.")
+        model_trace.compute_log_prob()
+        self.ordering = {name: frozenset(site["cond_indep_stack"])
+                         for name, site in model_trace.nodes.items()
+                         if site["type"] == "sample"}
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample" and not isinstance(site["fn"], _Subsample):
+                if is_validation_enabled():
+                    check_site_shape(site, self.max_iarange_nesting)
+                log_prob_shape = site["log_prob"].shape
+                self._enum_dims[site["name"]] = set((i for i in range(-len(log_prob_shape), -self.max_iarange_nesting)
+                                                     if log_prob_shape[i] > 1))
+
+    def _get_log_factors(self, model_trace):
+        """
+        Aggregates the `log_prob` terms into a list for each
+        ordinal.
+        """
+        model_trace.compute_log_prob()
+        log_probs = OrderedDict()
+        # Collect log prob terms per independence context.
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample" and not isinstance(site["fn"], _Subsample):
+                if is_validation_enabled():
+                    check_site_shape(site, self.max_iarange_nesting)
+                log_probs.setdefault(self.ordering[name], []).append(site["log_prob"])
+        return log_probs
+
+    def log_prob(self, model_trace):
+        """
+        Returns the log pdf of `model_trace` by appropriately handling
+        enumerated log prob factors.
+
+        :return: log pdf of the trace.
+        """
+        with shared_intermediates():
+            if not self.has_enumerable_sites:
+                return model_trace.log_prob_sum()
+            log_probs = self._get_log_factors(model_trace)
+            sum_dims = {site["log_prob"]: self._enum_dims[name] for name, site in model_trace.nodes.items()
+                        if site["type"] == "sample" and not isinstance(site["fn"], _Subsample)}
+            return contract_to_tensor(log_probs, sum_dims, frozenset()).sum()
```

Comment:
I'm surprised the `.sum()` is needed. Does #1405 fix this?

Reply:
The issue is that the tensor returned is not a scalar and has singleton dimensions (I think the same as the number of `iarange`s in the model). Let me check with #1405.

Reply:
So this is fixed in #1405, we wouldn't need the `.sum()` with that.